Implement Elastic Transcoder Support

indiv0 opened the pull request:
I'm interested in implementing Elastic Transcoder support (as I need it for a personal project), so I've begun a PR for it.

I'll post updates and questions here as I work on it.

# Current Status

* [x] Add `ets` feature to `Cargo.toml`
* [x] Generate `elastictranscoder` `Service` in `build.rs`
* [x] Implement `rest-json` generator
    * [x] Add `rest-json` generator
    * [x] Add `request_uri` format code generator function
    * [x] Add `querystring` format code generator function
    * [x] Manually verify it works for all `Ets` operations
        * [x] `cancel_job`
        * [x] `create_job`
        * [x] `create_pipeline`
        * [x] `create_preset`
        * [x] `delete_pipeline`
        * [x] `delete_preset`
        * [x] `list_jobs_by_pipeline`
        * [x] `list_jobs_by_status`
        * [x] `list_pipelines`
        * [x] `list_presets`
        * [x] `read_job`
        * [x] `read_pipeline`
        * [x] `read_preset`
        * [x] `test_role`
        * [x] `update_pipeline`
        * [x] `update_pipeline_notifications`
        * [x] `update_pipeline_status`
* [x] Add tests
    * `EtsClient`
        * generated functions
            * [ ] ~~`cancel_job`~~ (requires `IAM`)
            * [ ] ~~`create_job`~~ (requires `IAM`)
            * [ ] ~~`create_pipeline`~~ (requires `IAM`)
            * [x] `create_preset`
            * [ ] ~~`delete_pipeline`~~ (requires `IAM`)
            * [x] `delete_preset`
            * [ ] ~~`list_jobs_by_pipeline`~~ (requires `IAM`)
            * [x] `list_jobs_by_status` (could be improved)
                * once `create_job` is possible this should be modified to create a job then check for it in the list
            * [x] `list_pipelines` (could be improved)
                * once `create_pipeline` is possible this should be modified to create a pipeline then check for it in the list
            * [x] `list_presets`
            * [ ] ~~`read_job`~~ (requires `IAM`)
            * [ ] ~~`read_pipeline`~~ (requires `IAM`)
            * [x] `read_preset`
            * [ ] ~~`test_role`~~ (requires `IAM`)
            * [ ] ~~`update_pipeline`~~ (requires `IAM`)
            * [ ] ~~`update_pipeline_notifications`~~ (requires `IAM`)
            * [ ] ~~`update_pipeline_status`~~ (requires `IAM`)
        * extra tests
            * [x] `create_pipeline_without_arn`
* [ ] ~~Add documentation~~
* [x] Update `README.md`
    * [x] Add `ETS` to supported services list
* [ ] ~~Bump version~~
* [x] Squash commits
* [x] Rebase on master
* [ ] Update documentation on gh-pages

# Notes

* `HttpError`
    * `code` is now `Option<String>` because `HttpError`s for `Ecs` don't specify it
* `Metadata`
    * `service_abbreviation` is now `Option<String>` because `Ecs` doesn't specify it
* `generator`
    * `generate_protocol` now has a match statement to retrieve the `service_abbreviation` since it's now an `Option<String>`
* `HttpRequest`
    * `response_code` is a new field to accommodate the `responseCode` to expect, which is provided for certain `Ets` operations.

# Future PRs

* Improve error handling by expanding `AWSError` to support AWS error codes
* Add support for request retrying upon a "request throttled" response
* Add `IAM` support
* Add remaining `ETS` test cases (requires `IAM` support)

# Resources

* AWS
    * [Create Job - Amazon Elastic Transcoder][create-job]
    * [Read Job - Amazon Elastic Transcoder][read-job]
    * [Making HTTP Requests to Elastic Transcoder][making-http-requests]
    * [AWS CLI Reference - Elastic Transcoder][aws-ets-reference]
* botocore
    * [botocore - 2012-09-25/service-2.json][botocore-ets]
    * [botocore - serialize.py][botocore-serialize]

[aws-ets-reference]: http://docs.aws.amazon.com/cli/latest/reference/elastictranscoder/index.html
[botocore-ets]: https://github.com/boto/botocore/blob/develop/botocore/data/elastictranscoder/2012-09-25/service-2.json
[botocore-serialize]: https://github.com/boto/botocore/blob/master/botocore/serialize.py
[create-job]: http://docs.aws.amazon.com/elastictranscoder/latest/developerguide/create-job.html
[making-http-requests]: http://docs.aws.amazon.com/elastictranscoder/latest/developerguide/making-http-requests.html
[read-job]: http://docs.aws.amazon.com/elastictranscoder/latest/developerguide/get-job.html

indiv0 commented:
@jimmycuadra @matthewkmayer right now I have two questions:

1. Is the way I handle [`service_abbreviation` retrieval](https://github.com/rusoto/rusoto/pull/212/files#diff-b042b5677f883b9007470c2820b91c73R68) in `generate_protocol` acceptable?
It could be possibly refactored out to a function.
If you have an idea as to how to handle it without a `panic!`, I think that would be good too.

2. Do you have any idea why the request to Elastic Transcoder would hang (for about 20 seconds, then end with a `500` `Internal Server Error`) if I issue a `read_job` (see the error at the bottom)?

I've narrowed it down to an issue with the request payload.
If [`request.set_payload(Some(encoded.as_bytes()));`](https://github.com/rusoto/rusoto/pull/212/files#diff-51fd4ca8ee52b44a7d2879077e03a4abR28) is commented, it completes successfully.
I believe this has something to do with the `content-length` header (perhaps it's larger than it should be, so AWS waits for the remaining payload until timeout?).

    Response {
      status: InternalServerError,
      headers: Headers {
        Date: Thu, 07 Apr 2016 05:36:41 GMT,
        Transfer-Encoding: chunked, nnCoection: close,
      },
      version: Http11,
      url: Url {
        scheme: "https",
        scheme_data: Relative(RelativeSchemeData {
          username: "",
          password: None,
          host: Domain("elastictranscoder.us-east-1.amazonaws.com"),
          port: None, default_port: Some(443), path: ["2012-09-25", "jobs", "vLmPDjeioMmU"] }),
          query: None,
          fragment: None
        },
      status_raw: RawStatus(500, "Internal Server Error"),
      message: Http11Message {
        method: None,
        stream: Wrapper { obj: Some(Reading(ChunkedReader(chunk_remaining=unknown))) }
      }
    }

indiv0 commented:
So I've reworked `rest_json.rs` with the latest commit, and now payload loading is optional, conditional on there being no format arguments in the `request_uri`.
I think this is the proper solution because by my understanding of [`service-2.json`](https://github.com/boto/botocore/blob/develop/botocore/data/elastictranscoder/2012-09-25/service-2.json) a `rest-json` request to AWS will only have a payload if there are no `Member`s with a `location` field (i.e., a `Member` which is used to either format the URI or query string).

I've tested this with `create_job` and `read_job` and both appear to work (on the AWS side).

However, `create_job` returns an unknown [`AWSError`](http://pastebin.com/wPXKWRdp) when called, despite correctly creating the job in ETS.

indiv0 commented:
Actually, the [raw responses](http://pastebin.com/fWh5hL8V) for `create_job` look correct so I guess this makes this a JSON parsing error.
Will investigate further.

indiv0 commented:
Nope, I was wrong.

Manually deserializing the `CreateJobResponse` works as expected, so the issue isn't deserialization.

indiv0 commented:
Figured it out.
AWS is being "RESTful" and returning a `201 Created` status code, which results in `create_job` returning `Err(parse_json_protocol_error(&body))` since it expects a `200 OK`.

Fixing now.

indiv0 commented:
13f56474111e8cb203540e12571e02d5250a4cdf fixes the issue.

Both `create_job` and `read_job` now work as expected.

I have a new question now, since `s3` has the `S3Helper` struct, should I make a similar one implementing the methods of `Ets` which are confirmed working, or should I just mark them as such somewhere?

indiv0 commented:
1bfcfd3 should fix the compilation error on stable and beta, and 918d886 makes encoding of the `input` function parameter conditional on whether or not there's a payload (which fixes the warnings where it wasn't used).

matthewkmayer commented:
I'll be able to provide more input later, but to answer the question on the helpers structs: we're moving them all to the rusoto_helpers project, under the rusoto org on github.  https://github.com/rusoto/rusoto_helpers/pull/9/files has lots of changes but you can see the helpers in there.

Thanks for working on the ET service!

matthewkmayer commented on codegen/src/generator/mod.rs:
```diff
@@ -62,7 +65,15 @@ where P: GenerateProtocol {
+        service_name = match &service.metadata.service_abbreviation {
+            &Some(ref service_abbreviation) => service_abbreviation.as_str(),
+            &None => {
+                match service.metadata.endpoint_prefix {
```
I wonder if we should just switch to endpointprefix instead of service abbreviation for everything.  I think every service has that and may simplify this code.  Gotta check to see what other impacts that may have.

indiv0 replied:
@matthewkmayer do you want me to take a look at what impacts the change may have or are you going to handle it?

matthewkmayer replied:
I think it's worth making a new issue for and tackling outside of this PR.

indiv0 replied:
I've opened #220 to handle this.
I'll leave the way I've handled `service_name` as-is for now.

matthewkmayer commented on codegen/src/generator/rest_json.rs:
```diff
@@ -0,0 +1,144 @@
+    }
+}
+
+fn generate_uri_formatter(request_uri: &str, member_strings: &Vec<String>) -> String {
```
This section causes snake case warnings.  Elastictranscoder.rs codegen output for example:

```rust
let request_uri = format!("/2012-09-25/pipelines/{Id}/notifications" , Id = input.id);
```

```bash
warning: variable `__argId` should have a snake case name such as `__arg_id`, #[warn(non_snake_case)] on by default
```

We could massage both the request_uri field to replace `{Id}` with `{id`} and then call `.to_snake_case()` on member_strings on line 107.

indiv0 replied:
Alright, I'll take care of this.

indiv0 commented:
9000a8b fixes the snake case warnings.
I added a regex which searches for matches of the form `\{([\w\d]+)\}` in the `request_uri` and replaces the capture groups with a snake case version of the argument.

I added `lazy_static` as a dependency for its `lazy_static!` macro which allows a regex to be initialized lazily, when it is first called, instead of once for every time it is used.
I hope this is alright; if you prefer to avoid adding it as a dependency, I can initialize the regex in the `generate_methods` method, which means it'll be initialized once for each service which uses `rest-json` as its protocol.

`regex` was already a dependency of `rusoto-codegen`, but it wasn't called in `src/lib.rs` so I just added that.

indiv0 commented:
Yay a new syntex version!

The Travis CI builds have all failed due to syntex Registry mismatches.

indiv0 commented:
I've rebased on top of master now that #219 is merged, so the CI builds should run again, successfully this time.

indiv0 commented:
I'm currently working on adding support for `querystring` args, which are meant for the request params.

indiv0 commented:
With commit f8da841, `RestJsonGenerator` can now handle operations which (optionally) require query string parameters.

I've tested this change with `list_jobs_by_pipeline`, and it works correctly.

indiv0 commented:
0a032e7 fixes an issue where the payload was being loaded into the request despite there being params for the request preset.

139b076 adds initial tests for the `EtsClient`.

@matthewkmayer I've begun working on the test cases for the `EtsClient` methods.
I have two question with regards to this.

For the first: certain tasks such as `create_pipeline` require the creation of 2 `S3` buckets and 1 `IAM` role.
I could do this in the test case, but if the test fails, this leaves unnecessary `S3` buckets and `IAM` roles in the user's AWS services.
Preferably, I would use a cleanup method which ensures these items are removed, no matter whether or not the test fails.
It appears that the only way to do such a cleanup, currently, is with a framework such as [stainless][stainless] or [shiny][shiny].

Shiny doesn't look to be under active maintenance, but stainless appears to be.
I've investigated stainless for this use, but due to requiring compiler plugin support, it appears that stainless can only be used on Rust nightly.

So my questions are as follows:

1. Would it be acceptable to introduce the [stainless][stainless] framework as a `dev-dependency` and use it for test cases, keeping in mind that this will mean that Elastic Transcoder tests will only be able to run on Rust nightly?

If yes, I'll get to work on adding it.
If no, would you mind proposing an alternative?

2. Do the current test cases look OK? Is there anything you'd like changed about the way they're written?

[shiny]: https://github.com/farcaller/shiny
[stainless]: https://github.com/reem/stainless

matthewkmayer commented on codegen/Cargo.toml:
```diff
@@ -22,6 +22,7 @@ syntex = { version = "0.30.0", optional = true }
 
 [dependencies]
 Inflector = "0.2.0"
+lazy_static = "0.1"
```
For better or worse I'm a big fan of specifying as much of the version as possible.  Any complaints for locking this to 0.1.16?

indiv0 replied:
No complaints from me, I'll lock it to 0.1.16.

matthewkmayer commented:
I'm particularly against requiring nightly.  However I don't have a good solution for doing things otherwise - looks like the idea of `final` blocks hasn't quite landed in Rust.  Doing explicit, manual setup/teardown in each step isn't making me excited but may be the best option that fits my dislike for requiring nightly.

Current test cases look great.  Only thing I'd change is probably remove the comments - most are just restating what the code does.  I prefer more of "why" than "what" and think we could just remove them all in `tests/ets.rs`.

Anything I missed in open questions?

jimmycuadra commented:
I think requiring nightly is a non-starter. In the absence of setup/teardown hooks in libtest, you can sort of get by using a type that implements `Drop` to do cleanup, but unfortunately drop hooks are not guaranteed to run, so it's not foolproof. Check out the tests in [rust-etcd](https://github.com/jimmycuadra/rust-etcd) for an example of this approach, if you're interested in trying that route.

indiv0 commented:
@matthewkmayer alright I'll remove the extra comments (I kinda felt the same way about them).
Yes, that covers all my questions so far.

@jimmycuadra OK then I'll take a look at how rust-etcd handles tests and see if I can work that in.

jimmycuadra commented:
The SQS-specific parameters was not my choice. That was inherited from the previous version of the code. I agree that it should only matter what API protocol is being used, not which AWS service it's for. Go ahead and change them to use generic names!

indiv0 commented:
@jimmycuadra alright. Is it OK if I perform the name change in his PR or should it be done separately?

indiv0 commented:
Also, since the ETS test will need to perform tasks such as creating S3 buckets, how should that be handled, in terms of rusoto features?

Should the ETS tests explicitly require them with, for example:

    #![cfg(all(feature = "ets", feature = "s3", feature = "iam"))]

jimmycuadra commented:
Yep, this PR is fine (though I see you already did it!) Making all the relevant features a requirement for the tests seems reasonable to me.

indiv0 commented:
Yeah I prefer to make the changes then revert them if you guys say so than have to wait for a response everytime (non-blocking development :D).

Alright, I'll add the requirements.

Additionally, I have another question.
While researching cleanup for tests I found a [question](http://stackoverflow.com/questions/29963449/golang-like-defer-in-rust) you posted on StackOverflow.

I particularly like huon's answer, implementing a `defer!` macro.
I think this would work great for our use case.
Would you be opposed to me adding it to the code and using that for cleanup?

jimmycuadra commented:
If it makes the tests significantly easier to understand, I think it'd be okay to use scopeguard or the defer macro. That said, in rust-etcd I was able to achieve what I was asking in that question by just creating the `TestClient` wrapper type and implementing `Drop` and `Defer`, and the tests there are plenty clear for my liking. In other words, we need to think about the trade off between adding an extra dependency or having a lot of support code (for the defer macro) and the ergonomics of the test code.

indiv0 commented:
The problem with using `Drop` is the fact that the initialization/cleanup code is different for each test case and it seems wasteful to me to create and destroy S3 buckets, IAM roles, etc. for each test, even if they're not needed.

indiv0 commented:
Actually I just thought of an idea.
I could have the relevant items be of the form `Option<Bucket>` in the `TestEtsClient` struct (for example), then initialize them as necessary in the test, passing them to the `TestEtsClient` struct.

The `Drop` impl could then check if the fields are `None` or `Some`, and call the cleanup code if they're `Some`.

indiv0 commented:
Welp. I can't actually write the test case for `create_pipeline` because I need to be able to programmatically create and destroy an IAM role, and rusoto doesn't yet have IAM support.

I guess I'll write the test cases for the ones that we actually can support right now.

indiv0 commented:
On another note, clean up using `Option` types and a `Drop` impl works correctly!
Once I find a test which I can actually implement (i.e. doesn't rely on an unsupported service) and requires this, I'll push the changes.

indiv0 commented:
So I added an example test case and the `Drop` impl code.
Take a look at it, let me know what you think.

Also, I took a look through the list of tests (you can see it in the first post of this PR).
Most of them either directly require `IAM` support (such as `create_pipeline`) or transitively require `IAM` because they need another function to be supported (e.g. `create_job` needs `create_pipeline` which needs `IAM` support).

Now, this doesn't mean that the services won't work.
Instead of programmatically obtaining a `Role`, you could manually specify it as a `String` in the code.
We just have no way of writing tests for these functions until `IAM` support lands.

I'll write the test cases for the ones I can and proceed from there.

indiv0 commented on tests/ets.rs:
```diff
@@ -1,13 +1,110 @@
+                    Err(e) => { error!("Failed to delete S3 bucket: {}", e) }
+                };
+            });
+        });
```
Actually I think `if let` blocks might be better suited for this than `map`ping a closure over an `Option`.
Thoughts?

jimmycuadra replied:
Whichever you think is clearer. It's fine either way.

indiv0 replied:
I'll leave it as is since `if let` blocks wouldn't really make this shorter and I think that writing things functionally where possible is more "rustic".

indiv0 commented:
Hahahahahahaha!
I was getting intermittent failures when running the tests.
Specifically, `delete_job` was failing sometimes.
The error message wasn't too descriptive (just saying that it failed), and it was only erroring about 20% of the time I ran the test suite.

After turning on the debug logging, I got this gem:

    DEBUG:rusoto::ets: Response body: {"message":"Your application is submitting requests to Amazon Elastic Transcoder faster than the maximum request rate."}
    DEBUG:rusoto::ets: Response status: 400

Looking at the AWS documentation on [ETS rate limits](http://docs.aws.amazon.com/elastictranscoder/latest/developerguide/limits.html), the relevant portion is:

    Create Job: You can submit two Create Job requests per second per AWS account at a sustained rate; brief bursts of 100 requests per second are allowed.

I guess I was creating more than 2 jobs per second.
What should I do here?
Adding a timer in the tests doesn't feel like an ideal solution because this'll slow down their execution.
Plus, this doesn't fix the issue of several tests attempting to perform the same task simultaneously, which a delay won't fix.

indiv0 commented:
The way [`boto` handles it](https://github.com/danielgtaylor/botocore/commit/9494189032b8171bbf4385c5a87ba5f3c3a17eb9) is by checking the response headers for when `x-amzn-errortype` is equal to `ThrottlingException`.
They then retry the request.

Shouldn't be too difficult to implement a similar system.

indiv0 commented:
@jimmycuadra, @matthewkmayer, 9d9d626 completes the possible tests I can write for the `EtsClient` functions. The rest require `IAM`.

I'm currently testing the remaining functions to see if they work manually.

Once that's done, what should I move onto?

indiv0 commented:
I've finished verifying all of the non-testable functions manually.
They all work correctly.
I guess this means that the generator for `rest-json` is feature complete (for Elastic Transcoder support, that is).

indiv0 commented:
Alright, I've squashed the commits into separate logical chunks (I'll have to squash again to clean up the commit messages prior to merging).

Awaiting your review of the code and answers to my questions.

jimmycuadra commented:
The code looks good to me as is. I'd squash it all down to one commit. Thanks so much for this contribution!

indiv0 commented:
You're very welcome!

I've squashed it down. Should be good to merge now :)

jimmycuadra commented:
I'll wait for Matthew to do another look through and he can merge it if he doesn't find anything else to change.

matthewkmayer commented:
Totally getting on this in a few minutes.  😄

matthewkmayer commented:
Thanks for the PR, :shipit: !

We'll need an issue for adding IAM support, I'll make that presently.
